store/sqlstore/upgrade.go

```diff
@@ -1208,6 +1209,201 @@ func upgradeDatabaseToVersion537(sqlStore *SqlStore) {
+// 2. Marks threads as read for users where the last reply time of the thread
+//    is earlier than the time the user viewed the channel.
+//    Marking a thread means setting the mention count to zero and
+//    setting the last viewed at time of the the thread as the last viewed at time
+//    of the channel
+func fixCRTThreadCountsAndUnreads(sqlStore *SqlStore) {
+	countsAndLastReplyCTE := `
+		SELECT RootId, COUNT(Id) AS FixedReplyCount, max(Posts.CreateAt) as FixedLastReply
+		FROM Posts where Posts.DeleteAt = 0
+		GROUP BY RootId
+	`
+	updateReplyCountsQuery := `
+		WITH q AS (` + countsAndLastReplyCTE + `)
+		UPDATE Threads
+		SET ReplyCount = q.FixedReplyCount, LastReplyAt = q.FixedLastReply
+		FROM q
+		WHERE Threads.PostId = q.RootId;
+	`
+
+	if sqlStore.DriverName() == model.DATABASE_DRIVER_MYSQL {
+		updateReplyCountsQuery = `
+			UPDATE Threads
+			INNER JOIN (` + countsAndLastReplyCTE + `) as q
+			ON Threads.PostId = q.RootId
+			SET ReplyCount = q.FixedReplyCount, LastReplyAt = q.FixedLastReply
+		`
+	}
+
+	participantsCTE := `
+		SELECT COALESCE(NULLIF(RootId,''), Id) AS ThreadId, ARRAY_AGG(DISTINCT UserId) FixedParticipants
+		FROM Posts
+		WHERE Posts.DeleteAt = 0
+		GROUP BY ThreadId
+	`
+	updateParticipantsQuery := `
+		WITH q AS (` + participantsCTE + `)
+		UPDATE Threads
+		SET Participants = CONCAT('["', ARRAY_TO_STRING(q.FixedParticipants,'","'), '"]')
+		FROM q WHERE Threads.PostId = q.ThreadId
+	`
+	if sqlStore.DriverName() == model.DATABASE_DRIVER_MYSQL {
+		updateParticipantsQuery = `
+			UPDATE Threads
+			INNER JOIN (SELECT COALESCE(NULLIF(RootId,''), Id) AS ThreadId, GROUP_CONCAT(DISTINCT UserId SEPARATOR '\", \"') AS FixedParticipants FROM Posts WHERE Posts.DeleteAt = 0 GROUP BY ThreadId) as q
+			ON Threads.PostId = q.ThreadId
+			SET Participants = CONCAT('[\"', q.FixedParticipants, '\"]')
+                `
+	}
+
+	now := fmt.Sprintf("%d", model.GetMillis())
+	threadMembershipsCTE := `
+		SELECT PostId, UserId, ChannelMembers.LastViewedAt as CM_LastViewedAt, Threads.LastReplyAt
+		FROM Threads
+		INNER JOIN ChannelMembers on ChannelMembers.ChannelId = Threads.ChannelId
+		WHERE Threads.LastReplyAt <= ChannelMembers.LastViewedAt
+	`
+	updateThreadMembershipQuery := `
+		WITH q as (` + threadMembershipsCTE + `)
+		UPDATE ThreadMemberships set LastViewed = q.CM_LastViewedAt, UnreadMentions = 0, LastUpdated =` + now + `
+		FROM q WHERE ThreadMemberships.Postid = q.PostId AND ThreadMemberships.UserId = q.UserId
+	`
+	if sqlStore.DriverName() == model.DATABASE_DRIVER_MYSQL {
+		updateThreadMembershipQuery = `
+			UPDATE ThreadMemberships
+			INNER JOIN (` + threadMembershipsCTE + `) as q
+			ON ThreadMemberships.Postid = q.PostId AND ThreadMemberships.UserId = q.UserId
+			SET LastViewed = q.CM_LastViewedAt, UnreadMentions = 0, LastUpdated =` + now + `
+		`
+	}
+
+	if _, err := sqlStore.GetMaster().ExecNoTimeout(updateReplyCountsQuery); err != nil {
```

Comment:
Suggest to move the exec code to just below the SQL queries to keep it readable. 0/5.